zkvm/src/blockchain/block.rs

```diff
@@ -0,0 +1,121 @@
+use bulletproofs::BulletproofGens;
+use merlin::Transcript;
+
+use super::errors::BCError;
+use crate::{MerkleTree, Tx, TxID, TxLog, Verifier};
+
+#[derive(Clone, PartialEq)]
+pub struct BlockID(pub [u8; 32]);
+
+#[derive(Clone)]
+pub struct BlockHeader {
+    pub version: u64,
+    pub height: u64,
+    pub prev: BlockID,
+    pub timestamp_ms: u64,
+    pub txroot: [u8; 32],
+    pub utxoroot: [u8; 32],
+    pub nonceroot: [u8; 32],
+    pub refscount: u64,
+    pub ext: Box<[u8]>,
+}
+
+impl BlockHeader {
+    pub fn id(&self) -> BlockID {
+        let mut t = Transcript::new(b"ZkVM.blockheader");
+        t.commit_u64(b"version", self.version);
+        t.commit_u64(b"height", self.height);
+        t.commit_bytes(b"previd", &self.prev.0);
+        t.commit_u64(b"timestamp_ms", self.timestamp_ms);
+        t.commit_bytes(b"txroot", &self.txroot);
+        t.commit_bytes(b"utxoroot", &self.utxoroot);
+        t.commit_bytes(b"nonceroot", &self.nonceroot);
+        t.commit_u64(b"refscount", self.refscount);
+        t.commit_bytes(b"ext", &self.ext);
+
+        let mut result = [0u8; 32];
+        t.challenge_bytes(b"id", &mut result);
+        BlockID(result)
+    }
+
+    pub fn make_initial(timestamp_ms: u64, refscount: u64) -> BlockHeader {
+        BlockHeader {
+            version: 1,
+            height: 1,
+            prev: BlockID([0; 32]),
+            timestamp_ms: timestamp_ms,
+            txroot: [0; 32],
+            utxoroot: [0; 32],
+            nonceroot: [0; 32],
+            refscount: refscount,
+            ext: Box::new([]),
+        }
+    }
+}
+
+pub struct Block {
+    pub header: BlockHeader,
+    pub txs: Vec<Tx>,
+}
+
+impl Block {
+    pub fn validate(&self, prev: &BlockHeader) -> Result<Vec<TxLog>, BCError> {
+        if self.header.version < prev.version {
+            return Err(BCError::VersionReversion);
+        }
+        if self.header.version == 1 && self.header.ext.len() != 0 {
+            return Err(BCError::IllegalExtension);
+        }
+        if self.header.height != prev.height + 1 {
+            return Err(BCError::BadHeight);
+        }
+        if self.header.prev != prev.id() {
+            return Err(BCError::MismatchedPrev);
+        }
+        if self.header.timestamp_ms <= prev.timestamp_ms {
```

Comment:
There is a lot of conditions that are inverted and hard to read. How about having a helper method `check` (or `require`, or `ensure` or w/e) that uses the desired condition directly:

```rust
check(self.header.version >= prev.version, BCError::VersionReversion)?;
check(self.header.height == prev.height + 1, BCError::BadHeight)?;
check(self.header.prev == prev.id(), BCError::MismatchedPrev)?;
check(self.header.timestamp_ms > prev.timestamp_ms, BCError::BadBlockTimestamp)?;
...
```
 
Also: maybe it makes sense to factor out all BlockHeader rules into `BlockHeader::validate`, so headers can be validated w/o transactions, and we don’t have to repeat “self.header, self.header, self.header” many times?

Reply:
Much nicer, thanks!